src/libsyntax_ext/test_harness.rs

```diff
@@ -77,49 +72,47 @@ impl<'a> MutVisitor for TestHarnessGenerator<'a> {
     }
 
     fn flat_map_item(&mut self, i: P<ast::Item>) -> SmallVec<[P<ast::Item>; 1]> {
-        let ident = i.ident;
-        if ident.name != kw::Invalid {
-            self.cx.path.push(ident);
-        }
-        debug!("current path: {}", path_name_i(&self.cx.path));
-
         let mut item = i.into_inner();
         if is_test_case(&item) {
             debug!("this is a test item");
 
             let test = Test {
                 span: item.span,
-                path: self.cx.path.clone(),
+                ident: item.ident,
             };
-            self.cx.test_cases.push(test);
-            self.tests.push(item.ident);
+            self.tests.push(test);
         }
 
         // We don't want to recurse into anything other than mods, since
         // mods or tests inside of functions will break things
```

Comment:
Doesn't this PR solve this problem entirely?

Reply:
I think that it should work for modules inside expressions of we visited them. Is a bit more work for tests that are directly in blocks.

Reply:
Mostly, but it looks like some bits are still missing, see https://internals.rust-lang.org/t/allowing-test-on-inner-functions/11765.

Reply:
Ah that's what @matthewjasper was mentioning, i.e. `ast::Block`'s `NodeId` not being usable as the parent.